### ISSUE TITLE
"No connection pool for 'ApplicationRecord'" errors from ActiveRecord::Migration::CheckPending

### ISSUE THREAD

**Issue by jdelStrother** [rails/rails#43059]
### Steps to reproduce

* Checkout https://github.com/jdelStrother/migration-pool
* `bundle install`, `bin/rails db:reset`, `bin/rails server`
* Visit http://127.0.0.1:3000
* Check out main~1 (resulting in a change in db/migrate), and reload http://127.0.0.1:3000

### Expected behavior
The page refreshes without error, (or produces a message about having pending migrations)

### Actual behavior
This results in ActiveRecord::ConnectionNotEstablished - "No connection pool for 'ApplicationRecord'" found from the migration-checker middleware:
<details>
  <summary>Sample stacktrace</summary>

```
activerecord (6.1.5) lib/active_record/connection_adapters/abstract/connection_pool.rb:1125:in `retrieve_connection'
activerecord (6.1.5) lib/active_record/connection_handling.rb:327:in `retrieve_connection'
activerecord (6.1.5) lib/active_record/connection_handling.rb:283:in `connection'
activerecord (6.1.5) lib/active_record/model_schema.rb:380:in `table_exists?'
activerecord (6.1.5) lib/active_record/migration.rb:1108:in `get_all_versions'
activerecord (6.1.5) lib/active_record/migration.rb:1121:in `needs_migration?'
activerecord (6.1.5) lib/active_record/migration.rb:625:in `check_pending!'
activerecord (6.1.5) lib/active_record/migration.rb:590:in `block (2 levels) in call'
activesupport (6.1.5) lib/active_support/evented_file_update_checker.rb:59:in `execute'
activesupport (6.1.5) lib/active_support/evented_file_update_checker.rb:65:in `execute_if_updated'
activerecord (6.1.5) lib/active_record/migration.rb:597:in `block in call'
activerecord (6.1.5) lib/active_record/migration.rb:587:in `synchronize'
activerecord (6.1.5) lib/active_record/migration.rb:587:in `call'
actionpack (6.1.5) lib/action_dispatch/middleware/callbacks.rb:27:in `block in call'
activesupport (6.1.5) lib/active_support/callbacks.rb:98:in `run_callbacks'
actionpack (6.1.5) lib/action_dispatch/middleware/callbacks.rb:26:in `call'
actionpack (6.1.5) lib/action_dispatch/middleware/executor.rb:14:in `call'
actionpack (6.1.5) lib/action_dispatch/middleware/actionable_exceptions.rb:18:in `call'
actionpack (6.1.5) lib/action_dispatch/middleware/debug_exceptions.rb:29:in `call'
actionpack (6.1.5) lib/action_dispatch/middleware/show_exceptions.rb:33:in `call'
railties (6.1.5) lib/rails/rack/logger.rb:37:in `call_app'
railties (6.1.5) lib/rails/rack/logger.rb:26:in `block in call'
activesupport (6.1.5) lib/active_support/tagged_logging.rb:99:in `block in tagged'
activesupport (6.1.5) lib/active_support/tagged_logging.rb:37:in `tagged'
activesupport (6.1.5) lib/active_support/tagged_logging.rb:99:in `tagged'
railties (6.1.5) lib/rails/rack/logger.rb:26:in `call'
actionpack (6.1.5) lib/action_dispatch/middleware/remote_ip.rb:81:in `call'
actionpack (6.1.5) lib/action_dispatch/middleware/request_id.rb:26:in `call'
rack (2.2.3) lib/rack/method_override.rb:24:in `call'
rack (2.2.3) lib/rack/runtime.rb:22:in `call'
activesupport (6.1.5) lib/active_support/cache/strategy/local_cache_middleware.rb:29:in `call'
actionpack (6.1.5) lib/action_dispatch/middleware/executor.rb:14:in `call'
actionpack (6.1.5) lib/action_dispatch/middleware/static.rb:24:in `call'
rack (2.2.3) lib/rack/sendfile.rb:110:in `call'
actionpack (6.1.5) lib/action_dispatch/middleware/host_authorization.rb:148:in `call'
railties (6.1.5) lib/rails/engine.rb:539:in `call'
puma (5.4.0) lib/puma/configuration.rb:249:in `call'
puma (5.4.0) lib/puma/request.rb:77:in `block in handle_request'
puma (5.4.0) lib/puma/thread_pool.rb:340:in `with_force_shutdown'
puma (5.4.0) lib/puma/request.rb:76:in `handle_request'
puma (5.4.0) lib/puma/server.rb:440:in `process_client'
puma (5.4.0) lib/puma/thread_pool.rb:147:in `block in spawn_thread'
```

</details>

---

The problem appears to be that the migration connection has a reference to ApplicationRecord, but it's a old reference from before class reloading kicked in, so `self == ApplicationRecord` is no longer true, resulting in connection_specification_name returning `"ApplicationRecord"` rather than the expected `"ActiveRecord::Base"`

One potential fix that seems to work for me is to reimplement `ActiveRecord::Base.primary_class?` from this: https://github.com/rails/rails/blob/e7b0eddc70bf516bed56f892a40c57e8ddec559f/activerecord/lib/active_record/connection_handling.rb#L296-L298 to this:

```ruby
def primary_class?
  self == Base || self.name == "ApplicationRecord"
end
```

---

It looks like in edge-rails that's moved here: https://github.com/rails/rails/blob/e9f38a164e36104cc495d5ac5859e57f6be75f36/activerecord/lib/active_record/core.rb#L80-L88 but I think the logic still stands.

@eileencodes WDYT to changing that to something like the following?

```diff
def self.application_record_class? # :nodoc:
   if ActiveRecord.application_record_class
     self == ActiveRecord.application_record_class
   else
-    if defined?(ApplicationRecord) && self == ApplicationRecord
-      true
-    end
+    self.name == "ApplicationRecord"
   end
 end
```

[EDIT: Although, hmm... seems like you'd also need the `.name` comparison for the application_record_class]

### System configuration
**Rails version**: 6.1.5

**Ruby version**: 2.7.4

**Comment by eileencodes** [comment]
Hi @jdelStrother I tested this out on `main` and 6.1.5` and I can't reproduce what you're seeing. In all cases I tried I correctly see the check pending exception. Can you confirm if this is still happening for you? If you can, can you update the steps. 

Here's what I did:

1) Check out your app
2) `rake db:create`
3) `bin/rails s`
4) Go to browser, see check pending error
5) `git checkout main~1` with server running
6) Go to browser, reload, see check pending error.

**Comment by jdelStrother** [comment]
Hey @eileencodes, thanks for trying it out.
I can still reproduce it on 6.1.5 & 7.x, though I think your behaviour differs from mine because my original instructions were bad. I've updated them to include a `bin/rails db:reset` - could you give it one more try?

**Comment by eileencodes** [comment]
Ok I reproduced this but I don't have a fix at the moment. I have other work I'm exploring right now that might fix this but it's a part of a larger project and I don't have a timeline. In the meantime, a "fix" is to restart the server.

**Comment by eileencodes** [comment]
This was fixed in `main` by #45450.
